### PR TITLE
pkg/tls: Make it possible to load a CA from disk

### PR DESCRIPTION
**Pull request by olix0r**
In preparation for creating an Identity service that can chain off of an
existing CA, it's necessary to both (1) be able to create an
intermediate CA that can be used by the identity service and (2) be able
to load a CA from existing key material.

This changes the public API of the `tls` package to deal in actual key
types (rather than opaque blobs) and provides a set of helpers that can
be used to convert these credentials between common formats.

### CODE REVIEW

**Comment by olix0r** [comment]
This has been unit tested but needs more integration testing, especially with auto-inject...

Also, I'll probably take another pass at documenting the CA API... but wanted to share the general direction for early feedback.

File: pkg/tls/codec.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	if block.Type != "CERTIFICATE" {
+		return nil, nil, nil
+	}
+	c, err := x509.ParseCertificate(block.Bytes)
+	return c, crtb, err
+}
+
+// DecodeCrtPEM decodes PEM-encoded certificates from leaf to root.
+func DecodeCrtPEM(buf []byte) (crt Crt, err error) {
```
Comment (klingerf) [inline pkg/tls/codec.go]: Same comment here as above -- I'd return `*Crt`.

File: pkg/tls/ca.go [hunk 2]
```diff
@@ -29,14 +28,11 @@ type CA struct {
...
+	// PrivateKey of the CA.
+	PrivateKey *ecdsa.PrivateKey
 
-	// The CA's certificate.
-	root *x509.Certificate
-
-	// The PEM X.509 encoding of `root`
-	rootPEM string
+	// Crt is the certificate and trust chain.
+	Crt Crt
```
Comment (klingerf) [inline pkg/tls/ca.go]: Related to my comments in codec.go, I'd change this to:

```suggestion
	Crt *Crt
```

Note that if you do this you'll also have to update `newCA` to start initializing this field.

File: pkg/tls/ca.go [hunk 3]
```diff
@@ -49,97 +45,167 @@ type CA struct {
...
+	})
+	return &EndEntity{PrivateKey: pk, Crt: crt}, err
 }
 
-// createTemplate returns a certificate template for a non-CA certificate with
-// no subject name, no subjectAltNames. The template can then be modified into
-// a (root) CA template or an end-entity template by the caller.
+// SignEndEntity creates a new certificate that is valid for the
+// given DNS name, generating a new keypair for it.
+func (ca *CA) SignEndEntity(csr *x509.CertificateRequest) (Crt, error) {
```
Comment (klingerf) [inline pkg/tls/ca.go]: And finally I'd also have this func return `*Crt`, and then I'd embed `*Crt` in the `EndEntity` struct.

```diff
diff --git a/pkg/tls/ca.go b/pkg/tls/ca.go
index f157558f..c9f3cc5b 100644
--- a/pkg/tls/ca.go
+++ b/pkg/tls/ca.go
@@ -53,7 +53,7 @@ type Crt struct {
 
 type EndEntity struct {
 	PrivateKey *ecdsa.PrivateKey
-	Crt
+	*Crt
 }
 
 func newCA() *CA {
@@ -175,10 +175,10 @@ func (ca *CA) GenerateEndEntity(dnsName string) (*EndEntity, error) {
 
 // SignEndEntity creates a new certificate that is valid for the
 // given DNS name, generating a new keypair for it.
-func (ca *CA) SignEndEntity(csr *x509.CertificateRequest) (Crt, error) {
+func (ca *CA) SignEndEntity(csr *x509.CertificateRequest) (*Crt, error) {
 	pubkey, ok := csr.PublicKey.(*ecdsa.PublicKey)
 	if !ok {
-		return Crt{}, fmt.Errorf("CSR must contain an ECDSA public key: %+v", csr.PublicKey)
+		return nil, fmt.Errorf("CSR must contain an ECDSA public key: %+v", csr.PublicKey)
 	}
 
 	t := ca.createTemplate(pubkey)
@@ -193,16 +193,16 @@ func (ca *CA) SignEndEntity(csr *x509.CertificateRequest) (Crt, error) {
 
 	crtb, err := x509.CreateCertificate(rand.Reader, &t, ca.Crt.Certificate, pubkey, ca.PrivateKey)
 	if err != nil {
-		return Crt{}, fmt.Errorf("Failed to create certificate: %s", err)
+		return nil, fmt.Errorf("Failed to create certificate: %s", err)
 	}
 
 	crt, err := x509.ParseCertificate(crtb)
 	if err != nil {
-		return Crt{}, fmt.Errorf("Failed to parse certificate: %s", err)
+		return nil, fmt.Errorf("Failed to parse certificate: %s", err)
 	}
 
 	chain := append(ca.Crt.TrustChain, ca.Crt.Certificate)
-	return Crt{Certificate: crt, TrustChain: chain}, nil
+	return &Crt{Certificate: crt, TrustChain: chain}, nil
 }
 
 // createTemplate returns a certificate t for a non-CA certificate with
```

File: pkg/tls/codec.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	"fmt"
+	"io/ioutil"
+)
+
+func (crt *Crt) EncodePEM() []byte {
+	w := bytes.Buffer{}
+	n := len(crt.TrustChain)
+
+	// Serialize certificates from leaf to root.
+	_ = pem.Encode(&w, &pem.Block{Type: "CERTIFICATE", Bytes: crt.Certificate.Raw})
```
[the remark this answers is not in the document]
Comment (olix0r) [reply pkg/tls/codec.go]: these encodes basically have to fail (pem.EncodeToMemory doesn't expose an error type). the errors can only occur if the writer fails.. i'll at least add a comment to that effect, or perhaps a panic instead

File: pkg/tls/codec.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return DecodeTrustAnchorsPEM(b)
+}
+
+// ReadCrtPEM reads PEM-encoded certificates from the named file
+func ReadCrtPEM(path string) (Crt, error) {
```
[the remark this answers is not in the document]
Comment (olix0r) [reply pkg/tls/codec.go]: On my branch I've cleaned this up, but I still avoid using pointers in many places. I found this post to be informative about how to make decions https://stackoverflow.com/a/23551970

[the remark this answers is not in the document]
Comment (klingerf) [reply pkg/tls/codec.go]: Hmm, is it possible you haven't pushed this change back to origin? I don't see it in the PR.

File: pkg/tls/cred.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	// Cred is a container for a certificate, trust chain, and private key.
+	Cred struct {
+		PrivateKey *ecdsa.PrivateKey
+		Crt
+	}
+
+	// Crt is a container for a certificate and trust chain.
+	//
+	// The trust chain stores all issuer certificates from root at the head direct
+	// issuer at the tail.
```
Comment (klingerf) [inline pkg/tls/cred.go]: This might be more readable as:

```
	// The trust chain stores all issuer certificates from the root at the head to
	// the direct issuer at the tail.
```

File: pkg/tls/cred.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+		return nil, err
+	}
+	if !certificateMatchesKey(crt.Certificate, key) {
+		return nil, errors.New("tls: Public and private key do not match")
+	}
+	return &Cred{PrivateKey: key, Crt: *crt}, nil
+}
+
+// DecodePEMCrt decodes PEM-encoded certificates from leaf to root.
+func DecodePEMCrt(txt string) (*Crt, error) {
```
Comment (klingerf) [inline pkg/tls/cred.go]: Just noting that `ReadPEMCreds` and `DecodePEMCrt` are unused in this branch, but totally fine to keep them if they're intended for some future change.

File: controller/ca/controller.go [hunk 8]
```diff
@@ -111,7 +111,7 @@ func (c *CertificateController) syncNamespace(ns string) error {
 	configMap := &v1.ConfigMap{
 		ObjectMeta: metav1.ObjectMeta{Name: pkgK8s.TLSTrustAnchorConfigMapName},
 		Data: map[string]string{
-			pkgK8s.TLSTrustAnchorFileName: c.ca.TrustAnchorPEM(),
+			pkgK8s.TLSTrustAnchorFileName: c.ca.Cred.Crt.EncodeCertificatePEM(),
```
Comment (klingerf) [inline controller/ca/controller.go]: TIOLI since you're using embedded structs, you could shorten this to:

```suggestion
			pkgK8s.TLSTrustAnchorFileName: c.ca.Cred.EncodeCertificatePEM(),
```

File: controller/ca/controller.go [hunk 9]
```diff
@@ -141,16 +141,21 @@ func (c *CertificateController) syncSecret(key string) error {
...
 	if err != nil {
 		log.Errorf("Failed to issue certificate for %s", dnsName)
 		return err
 	}
 	secret := &v1.Secret{
 		ObjectMeta: metav1.ObjectMeta{Name: secretName},
 		Data: map[string][]byte{
-			pkgK8s.TLSCertFileName:       certAndPrivateKey.Certificate,
-			pkgK8s.TLSPrivateKeyFileName: certAndPrivateKey.PrivateKey,
+			pkgK8s.TLSCertFileName:       cred.Crt.Certificate.Raw,
```
Comment (klingerf) [inline controller/ca/controller.go]: Similar comment here:

```suggestion
			pkgK8s.TLSCertFileName:       cred.Certificate.Raw,
```

File: controller/proxy-injector/webhook_config.go [hunk 10]
```diff
@@ -27,14 +27,14 @@ type WebhookConfig struct {
 
 // NewWebhookConfig returns a new instance of initiator.
 func NewWebhookConfig(client kubernetes.Interface, controllerNamespace, webhookServiceName string, rootCA *tls.CA) (*WebhookConfig, error) {
-	trustAnchor := []byte(rootCA.TrustAnchorPEM())
+	trustAnchor := rootCA.Cred.Crt.EncodeCertificatePEM()
```
Comment (klingerf) [inline controller/proxy-injector/webhook_config.go]: And here:

```suggestion
	trustAnchor := rootCA.Cred.EncodeCertificatePEM()
```

File: pkg/tls/ca.go [hunk 11]
```diff
@@ -1,146 +1,216 @@
...
-	//
-	// Currently this is used for the CA's validity too, but nothing should
-	// assume that the CA's validity period is the same as issued certificates'
-	// validity.
-	validity time.Duration
+type (
+	// CA provides a certificate authority for TLS-enabled installs.
+	// Issuing certificates concurrently is not supported.
+	CA struct {
+		// cred contains the CA's credentials.
```
Comment (klingerf) [inline pkg/tls/ca.go]: Typo here: `cred` => `Cred`

File: pkg/tls/cred_test.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+	if err != nil {
+		t.Fatalf("Failed to decode PEM Crt: %s", err)
+	}
+
+	if err := crt.Verify(rootTrust, "endentity.test"); err != nil {
+		t.Fatal("Failed to verify round-tripped certificate")
+	}
+}
+
+func TestCredEncodeCeritificateAndTrustChain(t *testing.T) {
```
Comment (klingerf) [inline pkg/tls/cred_test.go]: Typo here: `Ceritificate` => `Certificate`

File: pkg/tls/cred.go [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+func (cred *Cred) EncodePrivateKeyP8() ([]byte, error) {
+	return x509.MarshalPKCS8PrivateKey(cred.PrivateKey)
+}
+
+// certificateMatchesKey returns whether the key and certificate match.
+func certificateMatchesKey(c *x509.Certificate, k *ecdsa.PrivateKey) (ok bool) {
+	pub, ok := c.PublicKey.(*ecdsa.PublicKey)
+	ok = ok && pub.X.Cmp(k.X) == 0 && pub.Y.Cmp(k.Y) == 0
+	return
+}
```
[the remark this answers is not in the document]
Comment (olix0r) [reply pkg/tls/cred.go]: ah yeah, i had that and then had to put some logging in there. will revert.

File: pkg/tls/cred.go (hunk 7 above)
Comment (olix0r) [reply pkg/tls/cred.go]: Yeah, they're used on my other branch...

File: controller/ca/controller.go (hunk 8 above)
Comment (olix0r) [reply controller/ca/controller.go]: omg this is so much better
